Bug Fix: fix the bug where plan solver misses the last point in multi-dimensional search

rwangsc18 opened the pull request:
## Description
Clearly describe changes in your PR, new feature or bug fixes.

Discover that our plan solver can miss the last point in the multi-dim iterator algorithm. It was found in the associated unit test, where the top return plan is supposed to have a score of 99 instead of 98. 

## Solution

Swap the sequence of iter.Next() and createPlanningSolutionCandidate(iter,) methods
Initialize the starting status in multi-dim iterator as [0, 0, -1] instead of [0, 0, 0]. 

## Testing
- [ ] Integration testing on Heroku staging
- [x] Updated unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

timwangmusic reviewed (commented):
Nice finding! Instead of setting the last state to -1, can we move iterative.Next() to the end of the for loop?

rwangsc18 commented:
> Nice finding! Instead of setting the last state to -1, can we move iterative.Next() to the end of the for loop?

Actually the call of iterator.Next() is after createPlanningSolutionCandidate(iter,) before this commit. Is it what you are asking about move `iterator.Next()` to the end of the loop? 

The main issue is that when the iterator points to the last possible candidate, the condition in `for iterator.HasNext()` is false.

timwangmusic commented:
> > Nice finding! Instead of setting the last state to -1, can we move iterative.Next() to the end of the for loop?
> 
> Actually the call of iterator.Next() is after createPlanningSolutionCandidate(iter,) before this commit. Is it what you are asking about move `iterator.Next()` to the end of the loop?
> 
> The main issue is that when the iterator points to the last possible candidate, the condition in `for iterator.HasNext()` is false, so the processing algorithm will miss the last candidate.

Yes, with `iterator.Next()` before `createPlanningSolutionCandidate` the first status used when creating candidate is `[0, 0, 0, ..., 0]`, right? So should it be equivalent to not changing iterator initialization and put the `iterator.Next()` after candidate creation?

rwangsc18 commented:
> > > Nice finding! Instead of setting the last state to -1, can we move iterative.Next() to the end of the for loop?
> > 
> > 
> > Actually the call of iterator.Next() is after createPlanningSolutionCandidate(iter,) before this commit. Is it what you are asking about move `iterator.Next()` to the end of the loop?
> > The main issue is that when the iterator points to the last possible candidate, the condition in `for iterator.HasNext()` is false, so the processing algorithm will miss the last candidate.
> 
> Yes, with `iterator.Next()` before `createPlanningSolutionCandidate` the first status used when creating candidate is `[0, 0, 0, ..., 0]`, right? So should it be equivalent to not changing iterator initialization and put the `iterator.Next()` after candidate creation?

Sg, I modified the iterator logics based on our discussion this week. I added a stop flag in the iterator struct, so we can have the iterator start from [0, 0, 0] and covers all possible points. Let me know if there is anything else to do for this PR.

timwangmusic reviewed (approved):
nice work! Elegant way to handle the stop flag.

rwangsc18 commented on planner/MDIter.go:
```diff
@@ -24,7 +26,9 @@ func (mdTagIter *MultiDimIterator) Init(categories []POI.PlaceCategory, placeClu
 	mdTagIter.Categories = categories
 	mdTagIter.Status = make([]int, len(mdTagIter.Categories))
 	mdTagIter.Size = make([]int, len(mdTagIter.Categories))
+	mdTagIter.Stop = false
```
Done.
